base_layer/key_manager/src/cipher_seed.rs

```diff
@@ -230,25 +243,29 @@ impl CipherSeed {
     }
 
     fn apply_stream_cipher(data: &mut Vec<u8>, passphrase: &str, salt: &[u8]) -> Result<(), KeyManagerError> {
-        let argon2 = Argon2::default();
         let blake2_nonce_hasher: VarBlake2b = VarBlake2b::new(size_of::<Nonce>())
             .expect("Should be able to create blake2 hasher; will only panic if output size is 0 or greater than 64");
 
         let mut encryption_nonce = [0u8; size_of::<Nonce>()];
+
+        // TODO: use hashing API to generate suitable domain separation instead
+        let label = b"chacha20_salt";
         blake2_nonce_hasher
+            .chain(label)
             .chain(salt)
             .finalize_variable(|res| encryption_nonce.copy_from_slice(res));
         let nonce_ga = Nonce::from_slice(&encryption_nonce);
 
-        // Create salt string stretched to the chacha nonce size, we only have space for 5 bytes of salt in the seed but
-        // will use key stretching to produce a longer nonce for the passphrase hash and the encryption nonce.
-        let salt_b64 = SaltString::b64_encode(&encryption_nonce)?;
+        // we take the last 32 bytes of the generated derived encryption key for Chacha20 cipher, see
+        // documentation
+        let derived_encryption_key: Vec<u8> = Self::generate_domain_separated_passphrase_hash(passphrase, salt)?
+            .into_iter()
+            .rev()
+            .take(32)
+            .rev()
+            .collect();
```

Comment:
Is there a more straightforward and/or more idiomatic way to do this?

Reply:
I refactored it to use slicing, it makes the code cleaner and simpler to read. Thanks !